core/src/main/scala/kafka/log/ProducerStateManager.scala

```diff
@@ -828,12 +834,41 @@ class ProducerStateManager(val topicPartition: TopicPartition,
    * Removes the producer state snapshot file metadata corresponding to the provided offset if it exists from this
    * ProducerStateManager, and deletes the backing snapshot file.
    */
-  private[log] def removeAndDeleteSnapshot(snapshotOffset: Long): Unit = {
+  private def removeAndDeleteSnapshot(snapshotOffset: Long): Unit = {
     Option(snapshots.remove(snapshotOffset)).foreach(_.deleteIfExists())
   }
+
+  /**
+   * Removes the producer state snapshot file metadata corresponding to the provided offset if it exists from this
+   * ProducerStateManager, and renames the backing snapshot file to have the Log.DeletionSuffix.
+   *
+   * Note: This method is safe to use with async deletes. If a race occurs and the snapshot file
+   *       is deleted without this ProducerStateManager instance knowing, the resulting exception on
+   *       SnapshotFile rename will be ignored and None will be returned.
+   */
+  private[log] def removeAndMarkSnapshotForDeletion(snapshotOffset: Long): Option[SnapshotFile] = {
+    Option(snapshots.remove(snapshotOffset)).flatMap { snapshot => {
+      // If the file cannot be renamed, it likely means that the file was deleted already.
+      // This can happen due to the way we construct an intermediate producer state manager
```

Comment:
I'm not entirely sure I understood this comment. Looking at the `LogLoader` code, it doesn't appear that we use the [intermediate](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/log/LogLoader.scala#L327) producer state manager to issue async deletions of snapshot files.

So, is it still possible that a missing file is a valid case?

Reply:
Agree with Kowshik. It's not clear if the comment is still valid.

Reply:
@junrao @kowshik I accidentally included the word `async` here, the deletion performed by the intermediate ProducerStateManager is done synchronously.

I'm referring to the case where we go through `LogLoader.recoverSegment`. We construct a new "intermediate" `ProducerStateManager` for segment recovery which is separate from the "real" `ProducerStateManager` captured in `LoadLogParams`. 

Segment recovery can use the intermediate `ProducerStateManager` to truncate snapshot files via `ProducerStateManager.truncateAndReload` in `Log.rebuildProducerState`. The `SnapshotFile` instances will be removed from the in-memory map for the "intermediate" `ProducerStateManager` in this case, but will remain for the "real" `ProducerStateManager` captured in `LoadLogParams`.

Reply:
@gardnervickers : Good point. It seems that LogLoader.recoverSegment() can both remove and add snapshots, both of which will be missing in the "real" ProducerStateManager captured in LoadLogParams. This can lead to the missing file issue you pointed out and also potentially cause LogLoad.load() to do an unnecessary expensive Log.rebuildProducerState().

@kowshik : I am wondering if we should let LoadLoader reload the snapshots in the "real" ProducerStateManager before calling Log.rebuildProducerState() in LogLoad.load().

Reply:
@gardnervickers : @kowshik mentioned that params.producerStateManager.removeStraySnapshots(params.segments.baseOffsets.toSeq)
in LogLoad.load() actually reloads the snapshots after log recovery. So, it seems that the issue you mentioned may not be a problem?

Reply:
@kowshik @junrao Right, we'll still end up with a a fully in-sync `ProducerStateManager` after `LogLoader.load(..)` runs. 

The problem can still occur though because we delete snapshots using both the "intermediate" and "real" `ProducerStateManager` prior to `removeStraySnapshots` at the end of `LogLoader.load`. 

1. `recoverSegment` can delete snapshots with the intermediate `ProducerStateManager`
2. `removeAndDeleteSegmentsAsync` will use the "real" `ProducerStateManager` to schedule async deletion. It may have a stale view of the present snapshots on the filesystem if #1 deleted snapshots, causing the rename to fail.
3. At the end of `LogLoader.load`, we will `removeStraySnapshots`, which will fix up any discrepancies between the contents of the log dir and the "real" `ProducerStateManager`.

Reply:
@gardnervickers : Thanks for the explanation. Makes sense.

@kowshik : The source of all the confusing is that we use the real PSM in some cases while using a temporary PSM in some other cases during recovery. The temporary PSM in recoverSegment() is used in 4 different places.

1. In recoverLog(). this is the case that we could just pass in the real PSM.
2. In completeSwapOperations(). We try to avoid recovering segment here in https://github.com/apache/kafka/pull/10763.
3 and 4. In loadSegmentFiles(). We probably need to clean this part of the logic a bit. If we are missing index file or the index file is corrupted, typically we can just rebuild the index without changing PSM. If the segment is truncated while rebuilding the index, we actually want to follow the process in step 1, by just removing the rest of the segments. So, we could also get rid of the temporary PSM in this case.

I am wondering if we could have a separate PR to get rid of the temporary PSM complete?

Reply:
@junrao That's a good point. Yes, we should get rid of the temporary PSM. I've created a jira tracking this improvement: https://issues.apache.org/jira/browse/KAFKA-12977. It is currently assigned to myself and I'll follow up on it.